doc/glusterfs.8
.\"  Copyright (c) 2008-2012 Red Hat, Inc. <http://www.redhat.com>
.\"  This file is part of GlusterFS.
.\"
.\"  This file is licensed to you under your choice of the GNU Lesser
.\"  General Public License, version 3 or any later version (LGPLv3 or
.\"  later), or the GNU General Public License, version 2 (GPLv2), in all
.\"  cases as published by the Free Software Foundation.
.\"
.\"
.\"
.TH GlusterFS 8 "Clustered File System" "07 March 2011" "Gluster Inc."
.SH NAME
GlusterFS \- clustered file system
.SH SYNOPSIS
.B glusterfs
.I [options] [mountpoint]
.PP
.SH DESCRIPTION
GlusterFS is a clustered file system, capable of scaling to several peta-bytes.
It aggregates various storage bricks over Infiniband RDMA or TCP/IP and
interconnect into one large parallel network file system. Storage bricks can
be made of any commodity hardware, such as x86-64 server with SATA-II RAID and
Infiniband HBA.

GlusterFS is fully POSIX compliant file system. On client side, it has dependency
on FUSE package, on server side, it works seemlessly on different operating systems.
Currently supported on GNU/Linux and Solaris.

.SH OPTIONS

.SS "Basic options"
.PP
.TP
\fB\-f, \fB\-\-volfile=VOLUME-FILE\fR
File to use as VOLUME-FILE.
.TP
\fB\-l, \fB\-\-log\-file=LOGFILE\fR
File to use for logging (the default is <INSTALL-DIR>/var/log/glusterfs/<MOUNT-POINT>.log).
.TP
\fB\-L, \fB\-\-log\-level=LOGLEVEL\fR
Logging severity.  Valid options are TRACE, DEBUG, INFO, WARNING, ERROR and CRITICAL (the default is INFO).
.TP
\fB\-s, \fB\-\-volfile\-server=SERVER\fR
Server to get the volume from.  This option overrides \fB\-\-volfile \fR option.
.TP
\fB\-\-volfile\-max\-fetch\-attempts=MAX\-ATTEMPTS\fR
Maximum number of connect attempts to server. This option should be provided with
\fB\-\-volfile\-server\fR option (the default is 1).

.SS "Advanced options"
.PP
.TP
\fB\-\-acl\fR
Mount the filesystem with POSIX ACL support.
.TP
\fB\-\-localtime\-logging\fR
Enable localtime log timestamps.
.TP
\fB\-\-debug\fR
Run in debug mode.  This option sets \fB\-\-no\-daemon\fR, \fB\-\-log\-level\fR to DEBUG,
and \fB\-\-log\-file\fR to console.
.TP
\fB\-\-enable\-ino32=BOOL\fR
Use 32-bit inodes when mounting to workaround application that doesn't support 64-bit inodes.
.TP
\fB\-\-fopen\-keep\-cache[=BOOL]\fR
Do not purge the cache on file open (default: false).
.TP
\fB\-\-mac\-compat=BOOL\fR
Provide stubs for attributes needed for seamless operation on Macs (the default is off).
.TP
\fB\-N, \fB\-\-no\-daemon\fR
Run in the foreground.
.TP
\fB\-p, \fB\-\-pid\-file=PIDFILE\fR
File to use as PID file.
.TP
\fB\-\-read\-only\fR
Mount the file system in 'read-only' mode.
.TP
\fB\-\-selinux\fR
Enable SELinux label (extended attributes) support on inodes.
.TP
\fB\-S, \fB\-\-socket\-file=SOCKFILE\fR
File to use as unix-socket.
.TP
\fB\-\-volfile\-id=KEY\fR
Key of the volume file to be fetched from the server.
.TP
\fB\-\-volfile\-server\-port=PORT\fR
Port number of volfile server.
.TP
\fB\-\-volfile\-server\-transport=TRANSPORT\fR
Transport type to get volume file from server (the default is tcp).
.TP
\fB\-\-volume\-name=VOLUME\-NAME\fR
Volume name to be used for MOUNT-POINT (the default is top most volume in VOLUME-FILE).
.TP
\fB\-\-worm\fR
Mount the filesystem in 'worm' mode.
.TP
\fB\-\-xlator\-option=VOLUME\-NAME.OPTION=VALUE\fR
Add/Override a translator option for a volume with the specified value.
.TP
\fB\-\-subdir\-mount=SUBDIR\-MOUNT\-PATH\fR
Mount subdirectory instead of the '/' of volume.

.SS "Fuse options"
.PP
.TP

\fB\-\-attr\-times\-granularity=NANOSECONDS\fR
Declare supported granularity of file attribute times (default is 0 which kernel handles as unspecified; valid real values are between 1 and 1000000000).
.TP
\fB\-\-attribute\-timeout=SECONDS\fR
Set attribute timeout to SECONDS for inodes in fuse kernel module (the default is 1).
.TP
\fB\-\-background\-qlen=N\fR
Set fuse module's background queue length to N (the default is 64).
.TP
\fB\-\-congestion\-threshold=N\fR
Set fuse module's congestion threshold to N (the default is 48).
.TP
\fB\-\-direct\-io\-mode=BOOL|auto\fR
Specify fuse direct I/O strategy (the default is auto).
.TP
\fB\-\-dump-fuse=PATH\fR
Dump fuse traffic to PATH
.TP
\fB\-\-entry\-timeout=SECONDS\fR
Set entry timeout to SECONDS in fuse kernel module (the default is 1).
.TP
\fB\-\-gid\-timeout=SECONDS\fR
Set auxiliary group list timeout to SECONDS for fuse translator (the default is 0).
.TP
\fB\-\-kernel-writeback-cache=BOOL\fR
Enable fuse in-kernel writeback cache.
.TP
\fB\-\-negative\-timeout=SECONDS\fR
Set negative timeout to SECONDS in fuse kernel module (the default is 0).
.TP
\fB\-\-auto\-invalidation=BOOL\fR
controls whether fuse-kernel can auto-invalidate attribute, dentry and
page-cache. Disable this only if same files/directories are not
accessed across two different mounts concurrently [default: on].
.TP
\fB\-\-volfile-check\fR
Enable strict volume file checking.

.SS "Miscellaneous Options"
.PP
.TP

\fB\-?, \fB\-\-help\fR
Display this help.
.TP
\fB\-\-usage\fR
Display a short usage message.
.TP
\fB\-V, \fB\-\-version\fR
Print the program version.

.PP
.SH FILES
/var/lib/glusterd/vols/*/*.vol
.SH EXAMPLES
mount a volume named foo on server bar with log level DEBUG on mount point
/mnt/foo

# glusterfs \-\-log\-level=DEBUG \-\-volfile\-id=foo \-\-volfile\-server=bar /mnt/foo

.SH SEE ALSO
.nf
\fBfusermount\fR(1), \fBmount.glusterfs\fR(8), \fBgluster\fR(8)
\fR
.fi
.SH COPYRIGHT
.nf
Copyright(c) 2006-2011  Red Hat, Inc.  <http://www.redhat.com>
\fR
.fi


NEWS
Gluster moves to Gerrit.


glusterfs-fuse/src/glusterfs.c
/*
  (C) 2006 Gluster core team <http://www.gluster.org/>
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License as
  published by the Free Software Foundation; either version 2 of
  the License, or (at your option) any later version.
    
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.
    
  You should have received a copy of the GNU General Public
  License along with this program; if not, write to the Free
  Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
  Boston, MA 02110-1301 USA
*/ 

#include <fuse.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/resource.h>
#include <netdb.h>
#include <argp.h>

#include "glusterfs-fops.h"
#include "logging.h"

/* using argp for command line parsing */
static char *mt_options = NULL;
static char *mount_point = NULL;
static char doc[] = "glusterfs is a glusterfs client";
static char argp_doc[] = "MOUNT-POINT";
static int cmd_def_log_level = GF_LOG_MAX;
static char *cmd_def_log_file = DEFAULT_LOG_FILE;
int gf_cmd_def_daemon_mode = GF_YES;


const char *argp_program_version = PACKAGE_NAME " " PACKAGE_VERSION;
const char *argp_program_bug_address = PACKAGE_BUGREPORT;

struct spec_location spec;
error_t parse_opts (int key, char *arg, struct argp_state *_state);

struct {
  char *f[2];
} f;

static struct argp_option options[] = {
  {"options", 'o', "OPTIONS", 0, "Filesystem mount options" },
  {"spec-file", 'f', "VOLUMESPEC-FILE", 0, "Load volume spec file VOLUMESPEC" },
  {"spec-server-ip", 's', "VOLUMESPEC-SERVERIP", 0, "Get volume spec file from VOLUMESPEC-SERVERIP"},
  {"spec-server-port", 'p', "VOLUMESPEC-SERVERPORT", 0, "connect to VOLUMESPEC_SERVERPORT on spec server"},
  {"log-level", 'L', "LOGLEVEL", 0, "Default LOGLEVEL"},
  {"log-file", 'l', "LOGFILE", 0, "Specify the file to redirect logs"},
  {"no-daemon", 'N', 0, 0, "Run glusterfs in foreground"},
  {"version", 'V', 0, 0, "print version information"},
  { 0, }
};
static struct argp argp = { options, parse_opts, argp_doc, doc };

static int
glusterfsd_print_version (void)
{
  printf ("%s\n", argp_program_version);
  printf ("Copyright (c) 2006 Z RESEARCH Inc. <http://www.zresearch.com>\n");
  printf ("GlusterFS comes with ABSOLUTELY NO WARRANTY.\nYou may redistribute copies of GlusterFS under the terms of the GNU General Public License.\n");
  exit (0);
}

error_t
parse_opts (int key, char *arg, struct argp_state *_state)
{
  switch (key){
  case 'f':
    spec.where = SPEC_LOCAL_FILE;
    spec.spec.file = strdup (arg);
    break;
  case 'o':
    mt_options = arg;
    break;
  case 's':
    spec.where = SPEC_REMOTE_FILE;
    spec.spec.server.ip = strdup (arg);
    break;
  case 'p':
    spec.spec.server.port = strdup (arg);
    break;
  case 'L':
    /* set log level */
    if (!strncmp (arg, "DEBUG", strlen ("DEBUG"))) {
      cmd_def_log_level = GF_LOG_DEBUG;
    } else if (!strncmp (arg, "NORMAL", strlen ("NORMAL"))) {
      cmd_def_log_level = GF_LOG_NORMAL;
    } else if (!strncmp (arg, "CRITICAL", strlen ("CRITICAL"))) {
      cmd_def_log_level = GF_LOG_CRITICAL;
    } else {
      cmd_def_log_level = GF_LOG_NORMAL;
    }
    break;
  case 'l':
    /* set log file */
    cmd_def_log_file = arg;
    break;
  case 'N':
    gf_cmd_def_daemon_mode = GF_NO;
    break;
  case 'V':
    glusterfsd_print_version ();
    break;
  case ARGP_KEY_NO_ARGS:
    argp_usage (_state);
    break;
  case ARGP_KEY_ARG:
    mount_point = arg;
    break;
  }
  return 0;
}
  
void 
args_init (int argc, char **argv)
{
  argp_parse (&argp, argc, argv, 0, 0, &f);
}


int
main (int argc, char *argv[])
{
  /* command line options: 
     -o allow_other -o default_permissions -o direct_io
  */

  struct rlimit lim;
  lim.rlim_cur = RLIM_INFINITY;
  lim.rlim_max = RLIM_INFINITY;
  setrlimit (RLIMIT_CORE, &lim);
  setrlimit (RLIMIT_NOFILE, &lim);

  args_init (argc, argv);
  if (gf_log_init (cmd_def_log_file) == -1) {
    fprintf (stderr, "%s: failed to open logfile \"%s\"\n", argv[0], cmd_def_log_file);
    return 1;
  }
  gf_log_set_loglevel (cmd_def_log_level);

  if (mount_point){
    return glusterfs_mount (&spec, mount_point, mt_options);
  } else{
    argp_help (&argp, stderr,ARGP_HELP_USAGE , argv[0]);
    return 1;
  }
}
